### DISCUSSION TITLE
Using the Interactivity API with Core Blocks

### DISCUSSION THREAD

**Discussion by dfilter** [WordPress/gutenberg#60479] [Interactivity API]
I'm wondering if it's possible currently to use the Interactivity API in conjunction with core blocks. 

As an example is there a way for me to use the core/button block and have it mutate the state of custom block (using the interactivity api)?

Thanks!

**Comment by michalczaplinski** [comment]
Yes, this is possible! You can use the [WP_HTML_Tag_Processor](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/) together with the Interactivity API to accomplish this. 

Mind you that you'd have to use the brand-new WordPress 6.5 to be able to use the Interactivity API.

Example which will log `"hello"` when clicking on **any** core Button block in the site:

```php
// some php file in your plugin

function my_custom_render( $block_content, $block ) {

	if ( $block['blockName'] !== 'core/button' ) {
		return $block_content;
	}
	
	// Enqueue the JS file. 
	// It assumes that the JS file is in the same directory as this PHP file
	wp_enqueue_script_module(
		'@my-plugin/entry',
		plugin_dir_url( __FILE__ ) . 'index.js',
	);
	
	$p = new WP_HTML_Tag_Processor( $block_content );
	$p->next_tag();
	
	// Make sure that the namesspace you pass as parameter ("my-plugin/button-interactivity") 
	// matches the namespace of the store in the JS file. 
	$p->set_attribute( 'data-wp-interactive', 'my-plugin/button-interactivity' );
	$p->set_attribute( 'data-wp-on--click', 'actions.logSomething' );
	return $p->get_updated_html();
}

add_filter( 'render_block', 'my_custom_render', 10, 2 );
```


And your JS file in the same folder as the PHP file:

```js
import { store } from '@wordpress/interactivity';

store("my-plugin/button-interactivity", {
  actions: {
    logSomething() {
      console.log('hello');
    }
  },
});
```

This example will work but the team working on the Interactivity API is thinking about ways to make this kind of use case even easier! There have been some explorations around "Behaviors UI" (https://github.com/WordPress/gutenberg/issues/50029) where the end goal is very similar to your use case but **more general**: _Being able to assign custom "behaviors" to common UI elements_. 

That work has been paused for now as the initial approach was not adequate but I hope that it can resume in the not so far future!

**Reply by dfilter** [reply]
Thanks so much @michalczaplinski it works beautifully!

**Comment by silaskoehler** [comment]
I’m working on an animation plugin for blocks that uses [motion.dev](https://motion.dev/) for the animation part. I want to make it compatible with the Interactivity API by using the init callback to trigger animations when a block appears.

My idea is to avoid creating a new custom block with all the animation settings. Instead, I want to integrate animations smoothly into core blocks and other existing blocks.

That said, I’m a bit worried about running into compatibility issues down the road. What’s the best practice here?

Also, I noticed in your example that the `render_block` filter assigns the namespace `my-plugin/button-interactivity` to the button block instead of the expected `core/button`. That got me wondering: what would happen if WordPress decided to add a default namespace for the button block in a future version? Could this lead to conflicts? And can a block have multiple namespaces?

Thanks for sharing your thoughts on this!

**Reply by luisherranz** [reply]
You don't need to rely on the default namespace, you can hardcode the namespace in the directive's value and add a unique ID in the directive's name (it can be the same namespace) to avoid conflicts.

```php
$p->set_attribute( 'data-wp-init--my-namespace', 'my-namespace::callbacks.initMotion' );
```

**Reply by silaskoehler** [reply]
Thanks, that was an helpful hint. 
Perhaps the documentation could include a separate section on namespaces and the `::`syntax.

**Reply by luisherranz** [reply]
Indeed, we have pending to write a section about namespaces and scopes in the `Core Concepts` section of the Interactivity API.

- https://github.com/WordPress/gutenberg/issues/62921
